src/mvba/abba/mod.rs

```diff
@@ -1,56 +1,577 @@
+                                }
+                            }
+                            _ => {
+                                return Err(Error::InvalidMessage(format!(
+                                    "invalid action. expected c_final, got {:?}",
+                                    c_final.action_str()
+                                )));
+                            }
+                        }
+
+                        // A weaker validity: an honest party may only decide on a value
+                        // for which it has the accompanying validating data.
+                        if action.value != PreVoteValue::One {
+                            return Err(Error::InvalidMessage(
+                                "initial value should be one".to_string(),
+                            ));
+                        }
+                    }
+                    PreVoteJustification::Hard(sig) => {
+                        // Hard pre-vote justification is the S-threshold signature for `(ID, pre-vote, r − 1, b)`
+                        let sign_bytes =
+                            self.pre_vote_bytes_to_sign(action.round - 1, &action.value)?;
+                        if !self.pub_key_set.public_key().verify(sig, &sign_bytes) {
+                            return Err(Error::InvalidMessage(
+                                "invalid hard-vote justification".to_string(),
+                            ));
+                        }
+                    }
+                    PreVoteJustification::Soft(sig) => {
+                        // Soft pre-vote justification is the S-threshold signature for `(ID, main-vote, r − 1, abstain)`
+                        let sign_bytes =
+                            self.main_vote_bytes_to_sign(self.r - 1, &MainVoteValue::Abstain)?;
+                        if !self.pub_key_set.public_key().verify(sig, &sign_bytes) {
+                            return Err(Error::InvalidMessage(
+                                "invalid soft-vote justification".to_string(),
+                            ));
+                        }
+                    }
+                }
+            }
+            Action::MainVote(action) => {
+                // check the validity of the S-signature share
+                let sign_bytes = self.main_vote_bytes_to_sign(action.round, &action.value)?;
+                if !self
+                    .pub_key_set
+                    .public_key_share(sender)
+                    .verify(&action.sig_share, &sign_bytes)
+                {
+                    return Err(Error::InvalidMessage("invalid signature share".to_string()));
+                }
+
+                match &action.justification {
+                    MainVoteJustification::NoAbstain(sig) => {
+                        // valid S-signature share on the message `(ID, pre-vote, r, b)`
+                        let sign_bytes =
+                            self.pre_vote_bytes_to_sign(action.round, &PreVoteValue::One)?;
```

Comment:
This main vote could be for zero, or it could be for one. The logic here currently hard codes to `one`
```suggestion
                        // valid S-signature share on the message `(ID, pre-vote, r, b)`
                        let sign_bytes =
                            self.pre_vote_bytes_to_sign(action.round, &action.value)?;
```